src/main/java/org/truffleruby/language/LazyRubyRootNode.java

```diff
@@ -34,32 +35,35 @@
 
 import java.util.List;
 
-public class LazyRubyRootNode extends RootNode implements InternalRootNode {
+public class LazyRubyRootNode extends RubyBaseRootNode implements InternalRootNode {
 
+    private final TruffleLanguage.ContextReference<RubyContext> contextReference;
+    private final SourceSection sourceSection;
     private final Source source;
     private final List<String> argumentNames;
 
-    @Child private Node findContextNode = RubyLanguage.INSTANCE.unprotectedCreateFindContextNode();
-
     @CompilationFinal private RubyContext cachedContext;
     @CompilationFinal private DynamicObject mainObject;
     @CompilationFinal private InternalMethod method;
 
     @Child private DirectCallNode callNode;
 
-    public LazyRubyRootNode(SourceSection sourceSection, FrameDescriptor frameDescriptor, Source source,
+    public LazyRubyRootNode(RubyLanguage language, SourceSection sourceSection, FrameDescriptor frameDescriptor, Source source,
                             List<String> argumentNames) {
-        super(RubyLanguage.class, sourceSection, frameDescriptor);
+        super(language, frameDescriptor);
+        contextReference = language.getContextReference();
+        this.sourceSection = sourceSection;
         this.source = source;
         this.argumentNames = argumentNames;
     }
 
     @Override
     public Object execute(VirtualFrame frame) {
-        final RubyContext context = RubyLanguage.INSTANCE.unprotectedFindContext(findContextNode);
+        final RubyContext context = contextReference.get();
 
         if (cachedContext == null) {
             CompilerDirectives.transferToInterpreterAndInvalidate();
+            context.ensureSetupFinished();
```

Comment:
I don't think he'll know what you are asking without any context.

In the new APIs, you cannot find the current context until the constructor of the context has finished. We run code during context creation. Therefore I've made the parts of context creation that need the context lazy, done when a lazy root node is first resolved.